### PR TITLE
[bitnami/kube-state-metrics] new chart

### PR DESCRIPTION
**Pull request by sameersbn**
**Description of the change**
kube-state-metrics is a simple service that listens to the Kubernetes API server and generates metrics about the state of the objects.

**Checklist** 
- [x] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[bitnami/chart]`)
- [x] If the chart contains a `values-production.yaml` apart from `values.yaml`, ensure that you implement the changes in both files

### CODE REVIEW

File: bitnami/kube-state-metrics/templates/_helpers.tpl [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+{{- define "kube-state-metrics.labels" -}}
+app.kubernetes.io/name: {{ include "kube-state-metrics.name" . }}
+helm.sh/chart: {{ include "kube-state-metrics.chart" . }}
+app.kubernetes.io/instance: {{ .Release.Name }}
+{{- if .Chart.AppVersion }}
+app.kubernetes.io/version: {{ .Chart.AppVersion | quote }}
+{{- end }}
+app.kubernetes.io/managed-by: {{ .Release.Service }}
+{{- if .Values.global.labels }}
+{{ toYaml .Values.global.labels }}
```
Comment (juan131) [inline bitnami/kube-state-metrics/templates/_helpers.tpl]: Suggestion:
```diff
-{{ toYaml .Values.global.labels }}
+{{- toYaml .Values.global.labels }}
```

File: bitnami/kube-state-metrics/templates/_helpers.tpl [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+{{- end -}}
+{{- end -}}
+
+{{/*
+matchLabels
+*/}}
+{{- define "kube-state-metrics.matchLabels" -}}
+app.kubernetes.io/name: {{ include "kube-state-metrics.name" . }}
+app.kubernetes.io/instance: {{ .Release.Name }}
+{{- end -}}
```
Comment (juan131) [inline bitnami/kube-state-metrics/templates/_helpers.tpl]: Add this macro after `kube-state-metrics.labels` macro. They're closely related

File: bitnami/kube-state-metrics/templates/_helpers.tpl [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+*/}}
+{{- define "kube-state-metrics.validateValues" -}}
+{{- $messages := list -}}
+{{- $messages := without $messages "" -}}
+{{- $message := join "\n" $messages -}}
+
+{{- if $message -}}
+{{- printf "\nVALUES VALIDATION:\n%s" $message | fail -}}
+{{- end -}}
+{{- end -}}
```
Comment (juan131) [inline bitnami/kube-state-metrics/templates/_helpers.tpl]: You didn't add any validation, was it intended?

File: bitnami/kube-state-metrics/templates/deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+      {{- end }}
+      {{- if .Values.securityContext.enabled }}
+      securityContext:
+        runAsUser: {{ .Values.securityContext.runAsUser }}
+        fsGroup: {{ .Values.securityContext.fsGroup }}
+      {{- end }}
+      hostNetwork: {{ .Values.hostNetwork }}
+      serviceAccountName: {{ template "kube-state-metrics.serviceAccountName" . }}
+      containers:
+        - name: {{ template "kube-state-metrics.name" . }}
```
Comment (juan131) [inline bitnami/kube-state-metrics/templates/deployment.yaml]: I would hardcode the name of the container to `kube-state-metrics`

File: bitnami/kube-state-metrics/templates/deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+  minReadySeconds: {{ .Values.minReadySeconds }}
+  template:
+    metadata:
+      {{- if .Values.podAnnotations }}
+      annotations: {{- include "kube-state-metrics.tplValue" ( dict "value" .Values.podAnnotations "context" $) | nindent 8 }}
+      {{- end }}
+      labels: {{- include "kube-state-metrics.labels" . | nindent 8 }}
+        {{- if .Values.podLabels }}
+{{ toYaml .Values.podLabels | indent 8 }}
+        {{- end }}
```
Comment (juan131) [inline bitnami/kube-state-metrics/templates/deployment.yaml]: Suggestion:
```diff
        {{- if .Values.podLabels }}
-{{ toYaml .Values.podLabels | indent 8 }}
+       {{- toYaml .Values.podLabels | nindent 8 }}
        {{- end }}
```

File: bitnami/kube-state-metrics/templates/deployment.yaml [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+          livenessProbe:
+            httpGet:
+              path: /healthz
+              port: http
+{{ toYaml .Values.livenessProbe | indent 12 }}
+          readinessProbe:
+            httpGet:
+              path: /
+              port: http
+{{ toYaml .Values.readinessProbe | indent 12 }}
```
Comment (juan131) [inline bitnami/kube-state-metrics/templates/deployment.yaml]: We should allow the users to disable the readiness/liveness probes.

File: bitnami/kube-state-metrics/templates/service.yaml [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  {{- if and (eq .Values.service.type "ClusterIP") .Values.service.clusterIP }}
+  clusterIP: {{ .Values.service.clusterIP }}
+  {{- end }}
+  ports:
+    - name: http
+      port: {{ .Values.service.port }}
+      targetPort: http
+      {{- if and .Values.service.nodePort (or (eq .Values.service.type "NodePort") (eq .Values.service.type "LoadBalancer")) }}
+      nodePort: {{ .Values.service.nodePort }}
+      {{- end }}
```
Comment (juan131) [inline bitnami/kube-state-metrics/templates/service.yaml]: Suggestion:
```diff
      {{- if and .Values.service.nodePort (or (eq .Values.service.type "NodePort") (eq .Values.service.type "LoadBalancer")) }}
      nodePort: {{ .Values.service.nodePort }}
+     {{- else if eq .Values.service.type "ClusterIP" }}
+     nodePort: null
      {{- end }}
```

File: bitnami/kube-state-metrics/templates/service.yaml [hunk 8]
```diff
@@ -0,0 +1,34 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ template "kube-state-metrics.fullname" . }}
+  annotations:
+    prometheus.io/scrape: "true"
+    {{- with .Values.service.annotations }}
+    {{- toYaml . | indent 4 }}
+    {{- end }}
```
Comment (juan131) [inline bitnami/kube-state-metrics/templates/service.yaml]: I would use `kube-state-metrics.tplValue` macho:
```diff
     {{- if .Values.service.annotations }}
-    {{- toYaml . | indent 4 }}
+    {{- include "kube-state-metrics.tplValue" (dict "value" Values.service.annotations "context" $) | nindent 4 }}
     {{- end }}
````

File: bitnami/kube-state-metrics/templates/service.yaml [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+  labels: {{- include "kube-state-metrics.labels" . | nindent 4 }}
+    {{- if .Values.service.labels }}
+{{ toYaml .Values.service.labels | indent 4 }}
+    {{- end }}
```
Comment (juan131) [inline bitnami/kube-state-metrics/templates/service.yaml]: Suggestion:
```diff
    {{- if .Values.service.labels }}
-{{ toYaml .Values.service.labels | indent 4 }}
+   {{- toYaml .Values.service.labels | nindent 4 }}
    {{- end }}
```

File: bitnami/kube-state-metrics/values-production.yaml [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+## Current available global Docker image parameters: imageRegistry and imagePullSecrets
+##
+global:
+  # imageRegistry: myRegistryName
+  # imagePullSecrets:
+  #   - myRegistryKeySecretName
+  # storageClass: myStorageClass
+
+  labels: {}
+  #   foo: bar
```
Comment (juan131) [inline bitnami/kube-state-metrics/values-production.yaml]: ??

File: bitnami/kube-state-metrics/templates/_helpers.tpl (hunk 3 above)
Comment (sameersbn) [reply bitnami/kube-state-metrics/templates/_helpers.tpl]: its a placeholder for any future validations

File: bitnami/kube-state-metrics/values-production.yaml (hunk 10 above)
Comment (sameersbn) [reply bitnami/kube-state-metrics/values-production.yaml]: added field description: `Additional labels to apply to all resources`

**Comment by sameersbn** [comment]
updated! please take another look

File: bitnami/kube-state-metrics/templates/_helpers.tpl (hunk 3 above)
Comment (juan131) [reply bitnami/kube-state-metrics/templates/_helpers.tpl]: All right!

File: bitnami/kube-state-metrics/templates/deployment.yaml [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+          ports:
+            - name: http
+              containerPort: 8080
+              protocol: TCP
+          {{- if .Values.livenessProbe.enabled }}
+          livenessProbe:
+            httpGet:
+              path: /healthz
+              port: http
+{{ toYaml .Values.livenessProbe | indent 12 }}
```
Comment (juan131) [inline bitnami/kube-state-metrics/templates/deployment.yaml]: You're including the `.Values.livenessProbe.enabled`, right?

**Pull request merged**